app/services/feeds/optimized.rb

```diff
@@ -0,0 +1,279 @@
+module Feeds
+  class Optimized
+    RANDOM_OFFSET_VALUES = [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 3, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13].freeze
+
+    def initialize(user: nil, number_of_articles: 35, page: 1, tag: nil)
+      @user = user
+      @number_of_articles = number_of_articles
+      @page = page
+      @tag = tag
+      @randomness = 3 # default number for randomly adjusting feed
+      @tag_weight = 1 # default weight tags play in rankings
+      @comment_weight = 0 # default weight comments play in rankings
+      @experience_level_weight = 1 # default weight for user experience level
+    end
+
+    def self.find_featured_story(stories)
+      featured_story =  if stories.is_a?(ActiveRecord::Relation)
+                          stories.where.not(main_image: nil).first
+                        else
+                          stories.detect { |story| !story.main_image.nil? }
+                        end
+      featured_story || Article.new
+    end
+
+    def find_featured_story(stories)
+      self.class.find_featured_story(stories)
+    end
+
+    def published_articles_by_tag
+      articles = Article.published.limited_column_select
+        .includes(top_comments: :user)
+        .page(@page).per(@number_of_articles)
+      articles = articles.cached_tagged_with(@tag) if @tag.present? # More efficient than tagged_with
+      articles
+    end
+
+    # Timeframe values from Timeframer::DATETIMES
+    def top_articles_by_timeframe(timeframe:)
+      published_articles_by_tag.where("published_at > ?", Timeframer.new(timeframe).datetime)
+        .order(score: :desc).page(@page).per(@number_of_articles)
+    end
+
+    def latest_feed
+      published_articles_by_tag.order(published_at: :desc)
+        .where("featured_number > ? AND score > ?", 1_449_999_999, -20)
```

Comment:
Where did we come up with `1_449_999_999` and `-20`? They feel a bit magic-numbery to me. Also, keeping in line with Forem generalization work, is this something that we want to define "higher up" as a constant (or maybe make configureable at some point?), the way that we do with `RANDOM_OFFSET_VALUES`?

Reply:
👍 for making them constants, ideally with descriptive names and comments if necessary.